File: _imports/pi-chart-claude/pi-chart/pi_chart/write.py
"""Write-side tool surface.

The only sanctioned way for pi-agent to mutate the chart. Enforces:
  - append-only (never modifies existing events.ndjson lines)
  - provenance required on every write (raises ValueError before append)
  - notes never silently overwrite existing files
  - id generation follows the evt_/note_ convention
  - recorded_at is always set to now at write time

Cross-process ID-counter robustness and immutable-input copying are deferred
to the TypeScript port; the in-memory counter here is sufficient for the
single-agent simulation.
"""

from __future__ import annotations

import json
import threading
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import yaml


_id_lock = threading.Lock()
_id_counter: dict[tuple[str, str], int] = {}  # (prefix, YYYYMMDDTHHMM) -> n


# Required envelope fields for every event regardless of type.
_REQUIRED_BASE = ("type", "subject", "effective_at", "author", "source", "status")

# Required additional fields for clinical claim types
# (mirrors schemas/event.schema.json conditional `allOf` block).
_CLINICAL_TYPES = {
    "observation", "assessment", "intent", "action", "communication", "artifact_ref",
}
_REQUIRED_CLINICAL = ("encounter_id", "certainty", "data", "links")


def _now_isoformat() -> str:
    return datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds")


def _day_dir(chart_root: Path, effective_at: str) -> Path:
    day = effective_at[:10]  # YYYY-MM-DD
    d = chart_root / "timeline" / day
    d.mkdir(parents=True, exist_ok=True)
    return d


def _check_provenance(event: dict) -> None:
    missing = [k for k in _REQUIRED_BASE if k not in event]
    if missing:
        raise ValueError(f"event missing required envelope fields: {missing}")
    if event.get("type") in _CLINICAL_TYPES:
        clinical_missing = [k for k in _REQUIRED_CLINICAL if k not in event]
        if clinical_missing:
            raise ValueError(
                f"clinical event ({event['type']}) missing required fields: {clinical_missing}"
            )


def next_event_id(effective_at: str | None = None) -> str:
    """Generate the next event id for a given effective-at timestamp.

    Format: evt_YYYYMMDDTHHMM_NN, where NN auto-increments within the minute.
    Not persistent across processes; the TS port replaces this with a
    file-probe pattern that survives restarts.
    """
    eff = effective_at or _now_isoformat()
    ymd_hm = eff.replace("-", "").replace(":", "")[:13]  # YYYYMMDDTHHMM
    key = ("evt", ymd_hm)
    with _id_lock:
        n = _id_counter.get(key, 0) + 1
        _id_counter[key] = n
    return f"evt_{ymd_hm}_{n:02d}"


def next_note_id(effective_at: str | None = None, slug: str | None = None) -> str:
    eff = effective_at or _now_isoformat()
    ymd_hm = eff.replace("-", "").replace(":", "")[:13]
    return f"note_{ymd_hm}" + (f"_{slug}" if slug else "")


def append_event(
    event: dict,
    *,
    chart_root: Path | str = ".",
) -> str:
    """Append a single event to the appropriate day's events.ndjson.

    Raises ValueError if required provenance fields are missing. Fills in
    `recorded_at` and `id` if absent. Returns the event id. Schema-level
    validation (e.g. enum membership) happens in `validate_chart` after a
    decision-cycle commits — this check is the cheaper, write-time guard.
    """
    _check_provenance(event)
    if "recorded_at" not in event:
        event["recorded_at"] = _now_isoformat()
    if "id" not in event:
        event["id"] = next_event_id(event.get("effective_at"))

    root = Path(chart_root)
    eff = event.get("effective_at") or event["recorded_at"]
    day = _day_dir(root, eff)
    events_path = day / "events.ndjson"
    with events_path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(event, separators=(",", ":")) + "\n")
    return event["id"]


def write_note(
    *,
    frontmatter: dict,
    body: str,
    chart_root: Path | str = ".",
    slug: str | None = None,
) -> Path:
    """Write a narrative note under timeline/YYYY-MM-DD/notes/HHMM_<slug>.md.

    Raises FileExistsError if the target path already exists — the chart is
    append-oriented and corrections are new notes that link `supersedes`,
    never overwrites. Fills in `recorded_at` and `id` if missing. Returns
    the written path.

    Caller is responsible for also appending a matching `communication` event
    in events.ndjson, OR (recommended) using `write_communication_note` which
    does both atomically.
    """
    root = Path(chart_root)
    fm = dict(frontmatter)
    if "recorded_at" not in fm:
        fm["recorded_at"] = _now_isoformat()
    eff = fm.get("effective_at") or fm["recorded_at"]
    if "id" not in fm:
        fm["id"] = next_note_id(eff, slug=slug)
    if "references" not in fm:
        fm["references"] = []

    day = _day_dir(root, eff)
    notes_dir = day / "notes"
    notes_dir.mkdir(exist_ok=True)
    hhmm = eff[11:16].replace(":", "")
    filename_slug = slug or fm["id"].split("_")[-1]
    filename = f"{hhmm}_{filename_slug}.md"
    path = notes_dir / filename

    if path.exists():
        raise FileExistsError(
            f"note already exists at {path} — chart is append-only; "
            f"write a new note that links supersedes/corrects to the prior id"
        )

    fm_yaml = yaml.safe_dump(fm, sort_keys=False, allow_unicode=True).strip()
    text = f"---\n{fm_yaml}\n---\n\n{body.strip()}\n"
    path.write_text(text, encoding="utf-8")
    return path


def write_communication_note(
    *,
    frontmatter: dict,
    body: str,
    communication_data: dict | None = None,
    chart_root: Path | str = ".",
    slug: str | None = None,
) -> tuple[Path, str]:
    """Write a note file AND append the matching communication event.

    Recommended over calling `write_note` + `append_event` separately because
    the inverse note ↔ communication link is easy to forget; the validator
    requires both directions.

    The communication event is constructed by mirroring the note's envelope
    (subject, encounter_id, effective_at, author, source, status) and setting
    `data.note_ref` to the note id. Pass extra communication-specific fields
    (audience, summary, subtype, ...) via `communication_data`.

    Returns (note_path, communication_event_id).
    """
    note_path = write_note(
        frontmatter=frontmatter,
        body=body,
        chart_root=chart_root,
        slug=slug,
    )
    fm = yaml.safe_load(note_path.read_text(encoding="utf-8").split("---", 2)[1])
    comm = {
        "type": "communication",
        "subtype": fm.get("subtype", "progress_note"),
        "subject": fm["subject"],
        "encounter_id": fm["encounter_id"],
        "effective_at": fm["effective_at"],
        "author": fm["author"],
        "source": fm["source"],
        "certainty": "performed",
        "status": fm.get("status", "final"),
        "data": {"note_ref": fm["id"], **(communication_data or {})},
        "links": {"supports": list(fm.get("references") or []), "supersedes": []},
    }
    event_id = append_event(comm, chart_root=chart_root)
    return note_path, event_id


def write_artifact_ref(
    *,
    artifact_path: Path | str,
    kind: str,
    description: str,
    encounter_id: str,
    subject: str,
    effective_at: str | None = None,
    chart_root: Path | str = ".",
    author: dict | None = None,
) -> str:
    """Register a binary artifact (image, PDF, waveform) as an artifact_ref event.

    The file itself should already live under `artifacts/`. This creates the
    pointer event in events.ndjson and returns the event id.
    """
    event = {
        "type": "artifact_ref",
        "subtype": kind,
        "subject": subject,
        "encounter_id": encounter_id,
        "effective_at": effective_at or _now_isoformat(),
        "author": author or {"id": "pi-agent", "role": "rn_agent"},
        "source": {"kind": "artifact_ingest"},
        "certainty": "observed",
        "status": "final",
        "data": {
            "kind": kind,
            "path": str(artifact_path),
            "description": description,
        },
        "links": {"supports": [], "supersedes": []},
    }
    return append_event(event, chart_root=chart_root)

File: _imports/pi-chart-ts/src/write.ts
// Write-side tool surface. The only sanctioned way to mutate the chart.
//
// Contracts:
//   - `appendEvent` requires base envelope fields, plus the four
//     clinical-only fields when type is observation/assessment/intent/
//     action/communication/artifact_ref. Throws on missing.
//   - Inputs are not mutated; the function works on a copy.
//   - `writeNote` refuses to overwrite an existing path. Corrections must
//     be new notes that link supersedes/corrects.
//   - `writeCommunicationNote` writes the note AND appends the matching
//     communication event so the bidirectional note ↔ comm validator
//     check passes by construction.
//   - `nextEventId` file-probes the day's events.ndjson for the max
//     suffix; safe across process restarts within a single writer.

import { promises as fs } from "node:fs";
import path from "node:path";
import {
  appendNdjsonLine,
  atomicWriteFile,
  ensureDir,
  fileExists,
  formatFrontmatterMarkdown,
  iterNdjson,
  parseFrontmatter,
} from "./fs-util.js";
import type {
  Author,
  ClinicalType,
  EventInput,
  NoteFrontmatter,
  NoteFrontmatterInput,
} from "./types.js";

const REQUIRED_BASE = [
  "type",
  "subject",
  "effective_at",
  "author",
  "source",
  "status",
] as const;

const CLINICAL_TYPES: ReadonlySet<ClinicalType> = new Set([
  "observation",
  "assessment",
  "intent",
  "action",
  "communication",
  "artifact_ref",
]);

const REQUIRED_CLINICAL = [
  "encounter_id",
  "certainty",
  "data",
  "links",
] as const;

function checkProvenance(ev: EventInput): void {
  const missing = REQUIRED_BASE.filter(
    (k) => (ev as any)[k] === undefined || (ev as any)[k] === null,
  );
  if (missing.length) {
    throw new Error(
      `event missing required envelope fields: ${missing.join(", ")}`,
    );
  }
  if (CLINICAL_TYPES.has(ev.type as ClinicalType)) {
    const cMissing = REQUIRED_CLINICAL.filter(
      (k) => (ev as any)[k] === undefined,
    );
    if (cMissing.length) {
      throw new Error(
        `clinical event (${ev.type}) missing required fields: ${cMissing.join(", ")}`,
      );
    }
  }
}

function nowIsoWithOffset(): string {
  // Local time with offset; matches Python's
  // `datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds")`.
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  const y = d.getFullYear();
  const M = pad(d.getMonth() + 1);
  const D = pad(d.getDate());
  const h = pad(d.getHours());
  const m = pad(d.getMinutes());
  const s = pad(d.getSeconds());
  const tzMin = -d.getTimezoneOffset();
  const sign = tzMin >= 0 ? "+" : "-";
  const tzH = pad(Math.floor(Math.abs(tzMin) / 60));
  const tzM = pad(Math.abs(tzMin) % 60);
  return `${y}-${M}-${D}T${h}:${m}:${s}${sign}${tzH}:${tzM}`;
}

function ymdHmFromIso(iso: string): string {
  // "2026-04-18T08:30:00-05:00" → "20260418T0830"
  const cleaned = iso.replace(/[-:]/g, "");
  return cleaned.slice(0, 13);
}

function dayDirFromIso(chartRoot: string, iso: string): string {
  return path.join(chartRoot, "timeline", iso.slice(0, 10));
}

/** Probe `chartRoot/timeline/<day>/events.ndjson` for max NN given ymdHm. */
export async function nextEventId(opts: {
  chartRoot: string;
  effectiveAt?: string;
}): Promise<string> {
  const eff = opts.effectiveAt ?? nowIsoWithOffset();
  const ymdHm = ymdHmFromIso(eff);
  const dayPath = dayDirFromIso(opts.chartRoot, eff);
  const eventsPath = path.join(dayPath, "events.ndjson");
  let max = 0;
  if (await fileExists(eventsPath)) {
    const re = new RegExp(`^evt_${ymdHm}_(\\d{2,})$`);
    for await (const [, ev] of iterNdjson(eventsPath)) {
      const id: unknown = ev?.id;
      if (typeof id !== "string") continue;
      const m = id.match(re);
      if (m) {
        const n = parseInt(m[1], 10);
        if (n > max) max = n;
      }
    }
  }
  return `evt_${ymdHm}_${String(max + 1).padStart(2, "0")}`;
}

export function nextNoteId(opts: {
  effectiveAt?: string;
  slug?: string;
}): string {
  const eff = opts.effectiveAt ?? nowIsoWithOffset();
  const ymdHm = ymdHmFromIso(eff);
  return `note_${ymdHm}` + (opts.slug ? `_${opts.slug}` : "");
}

export async function appendEvent(
  event: EventInput,
  opts: { chartRoot: string },
): Promise<string> {
  checkProvenance(event);
  // Immutable input: never mutate caller's object.
  const ev: any = { ...event };
  if (!ev.recorded_at) ev.recorded_at = nowIsoWithOffset();
  if (!ev.id) {
    ev.id = await nextEventId({
      chartRoot: opts.chartRoot,
      effectiveAt: ev.effective_at,
    });
  }
  const dayDir = dayDirFromIso(opts.chartRoot, ev.effective_at);
  await ensureDir(dayDir);
  await appendNdjsonLine(path.join(dayDir, "events.ndjson"), ev);
  return ev.id as string;
}

export interface WriteNoteOpts {
  frontmatter: NoteFrontmatterInput;
  body: string;
  chartRoot: string;
  slug?: string;
}

export async function writeNote(opts: WriteNoteOpts): Promise<string> {
  const fm: any = { ...opts.frontmatter };
  if (!fm.recorded_at) fm.recorded_at = nowIsoWithOffset();
  const eff: string = fm.effective_at ?? fm.recorded_at;
  if (!fm.id) fm.id = nextNoteId({ effectiveAt: eff, slug: opts.slug });
  if (!fm.references) fm.references = [];

  const dayDir = dayDirFromIso(opts.chartRoot, eff);
  await ensureDir(dayDir);
  const notesDir = path.join(dayDir, "notes");
  await ensureDir(notesDir);

  const hhmm = eff.slice(11, 16).replace(":", "");
  const idTail = (fm.id as string).split("_").pop() ?? "note";
  const filenameSlug = opts.slug ?? idTail;
  const filename = `${hhmm}_${filenameSlug}.md`;
  const target = path.join(notesDir, filename);

  if (await fileExists(target)) {
    throw new Error(
      `note already exists at ${target} — chart is append-only; ` +
        `write a new note that links supersedes/corrects to the prior id`,
    );
  }

  const content = formatFrontmatterMarkdown(fm, opts.body);
  await atomicWriteFile(target, content);
  return target;
}

export interface WriteCommunicationNoteOpts extends WriteNoteOpts {
  communicationData?: Record<string, unknown>;
}

export async function writeCommunicationNote(
  opts: WriteCommunicationNoteOpts,
): Promise<{ notePath: string; eventId: string }> {
  const notePath = await writeNote(opts);
  const text = await fs.readFile(notePath, "utf8");
  const [fm] = parseFrontmatter(text);
  const note = fm as NoteFrontmatter;

  const commEvent = {
    type: "communication" as const,
    subtype: note.subtype ?? "progress_note",
    subject: note.subject,
    encounter_id: note.encounter_id,
    effective_at: note.effective_at,
    author: note.author,
    source: note.source,
    certainty: "performed" as const,
    status: note.status ?? "final",
    data: { note_ref: note.id, ...(opts.communicationData ?? {}) },
    links: { supports: [...note.references], supersedes: [] },
  };
  const eventId = await appendEvent(commEvent, { chartRoot: opts.chartRoot });
  return { notePath, eventId };
}

export async function writeArtifactRef(opts: {
  artifactPath: string;
  kind: string;
  description: string;
  encounterId: string;
  subject: string;
  effectiveAt?: string;
  chartRoot: string;
  author?: Author;
}): Promise<string> {
  const event = {
    type: "artifact_ref" as const,
    subtype: opts.kind,
    subject: opts.subject,
    encounter_id: opts.encounterId,
    effective_at: opts.effectiveAt ?? nowIsoWithOffset(),
    author: opts.author ?? { id: "pi-agent", role: "rn_agent" },
    source: { kind: "artifact_ingest" },
    certainty: "observed" as const,
    status: "final" as const,
    data: {
      kind: opts.kind,
      path: opts.artifactPath,
      description: opts.description,
    },
    links: { supports: [], supersedes: [] },
  };
  return appendEvent(event, { chartRoot: opts.chartRoot });
}
